File: source/framework/api/python/test/lue_api_test/stress/local_operation.py
#!/usr/bin/env python
import numpy as np

import lue.framework_x as lfrx

from stress import main, Shape


class Model(lfrx.Model):
    def __init__(self, array_shape: Shape, partition_shape: Shape | None):
        super().__init__()
        self.array_shape = array_shape
        self.partition_shape = partition_shape

    def initialize(self):
        self.state = lfrx.uniform(
            self.array_shape,
            self.partition_shape,
            lfrx.as_field(lfrx.create_scalar(0, np.float32)),
            lfrx.as_field(lfrx.create_scalar(1, np.float32)),
        )

    def simulate(self, iteration: int):
        self.state = self.state + self.state - self.state + self.state - self.state

        return lfrx.as_state(self.state)


if __name__ == "__main__":
    main(Model)

File: source/framework/api/cxx/test/stress/local_operation.cpp
#include "stress.hpp"
#include "lue/framework/api/cxx.hpp"
#include "lue/framework/model.hpp"
#include "lue/framework.hpp"
#include <hpx/hpx_main.hpp>


namespace {

    class Model: public lue::Model
    {

        public:

            static lue::Rank const rank{2};

            using Shape = lue::Shape<lue::Count, rank>;

            using FloatElement = lue::FloatingPointElement<0>;


            Model(Shape const array_shape, Shape const partition_shape):
                lue::Model{},
                _array_shape{array_shape},
                _partition_shape{partition_shape}

            {
            }


            void initialize() final
            {
                using namespace lue::api;

                _field = uniform(_array_shape, _partition_shape, FloatElement{0}, FloatElement{1});
            }


            auto simulate([[maybe_unused]] lue::Count const time_step) -> hpx::shared_future<void> final
            {
                using namespace lue::api;

                _field = _field + _field - _field + _field - _field;

                return as_state(_field);
            }


        private:

            Shape _array_shape;

            Shape _partition_shape;

            lue::api::Field _field;
    };

}  // Anonymous namespace


auto main(int argc, char* argv[]) -> int
{
    return lue::stress::run_stress_test<Model>(argc, argv);
}
